BlocklyEvents (first pass, create events only, foundations of #309)

AnmAtAnm opened the pull request:
Event classes, listener hooks, and one listener implementation. DevTests activity now has a "Log Events" checkbox in it's action bar menu.  Create event works end-to-end.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-android/335)
<!-- Reviewable:end -->

AnmAtAnm commented:
PTAL @roboerikg 

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/335#-:-KL3ec0vL8-tCpU6MkaM:1290518790)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
General approach looks good. Some cleanup before submitting the first draft.

---

Review status: 0 of 10 files reviewed at latest revision, 22 unresolved discussions.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/DevTestsActivity.java, line 83 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHXqb0ywEcIAJsaTuf:-KLHXqb1Csu8P3OeQ9Qu:1953516138) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklydemo/src/main/java/com/google/blockly/android/demo/DevTestsActivity.java#L83)):*
> ```Java
>         int id = item.getItemId();
> 
>         if (item == mLogEventsMenuItem) {
> ```

Why aren't you using id == R.id.log_events_menuitem? Then you could toggle 'item' directly instead of keeping a reference to it and you can remove the onCreateOptionsMenu override.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java, line 70 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLH_B4O7w_4AXCL8PGa:-KLH_B4P9zFs3Vyg5KWI:-613717857) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java#L70)):*
> ```Java
>     private static final String SERIALIZED_WORKSPACE_KEY = "SERIALIZED_WORKSPACE";
> 
>     public interface Listener {
> ```

s/Listener/EventsCallback

Add Javadoc

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java, line 77 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLH_OsBdGNT6xS7jxZi:-KLH_OsBdGNT6xS7jxZj:-1526659728) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java#L77)):*
> ```Java
> 
>         /**
>          * Called when an event group contains at least on of the requested types.  Some events in
> ```

"Called when a group of events are fired and at least one event is of a type specified by {@link #getTypesBitmask}. All of the events are not guaranteed to be of a specified type."

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java, line 837 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHZ0KJM_05oV5w1eB1:-KLHZ0KJM_05oV5w1eB2:695833240) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java#L837)):*
> ```Java
>                 List<String> ids = new ArrayList<>();
>                 block.addAllBlockIds(ids);
>                 event = new BlocklyEvent.CreateEvent(
> ```

This should only be a create event if it's a new block, otherwise it's a move event with an old parent.

There are also cases where addRootBlock is called as part of a remove, please add a TODO to investigate.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 26 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHayjHviw0zbJn-2Xg:-KLHayjIw8gTtvVK0JGm:-1358787498) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L26)):*
> ```Java
> 
> /**
>  * Base class to all Blockly events.
> ```

s/to/for

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 39 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHb42a4ULVGxx0vmH_:-KLHb42a4ULVGxx0vmHa:1832524326) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L39)):*
> ```Java
>     public static final int TYPE_MOVE = 1 << 3;
>     public static final int TYPE_UI = 1 << 4;
>     // When adding an event type, update TYPE_ID_COUNT, TYPE_ALL, TYPE_ID_TO_NAME, and TYPE_NAME_TO_ID.
> ```

nit: >100 cols

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 92 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHbc5CfCFafR_06i5f:-KLHbc5CfCFafR_06i5g:-1716409538) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L92)):*
> ```Java
> 
>     /**
>      * Ensures {@code typeId} is a singular valid event it.
> ```

s/it/id

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 95 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHc3ozCvm1sa-gwqDN:-KLHc3ozCvm1sa-gwqDO:1117022097) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L95)):*
> ```Java
>      * @param typeId The typeId to test.
>      */
>     public static void validateEventType(int typeId) {
> ```

These validate methods can probably be private.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 172 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHcWIHTrHXDCzHlQDC:-KLHcWIIV8e1xtYCuC37:328940337) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L172)):*
> ```Java
>     protected String mGroupId;
> 
>     protected BlocklyEvent(@EventType int typeId, @Nullable String blockId,
> ```

javadoc

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 240 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHd15020ENc_nU03Fr:-KLHd15020ENc_nU03Fs:1236424627) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L240)):*
> ```Java
>      * Event fired when a block's field value has changed.
>      */
>     public static class ChangeEvent extends BlocklyEvent {
> ```

Consider making the concrete event classes final. It'll discourage breaking compatibility with other events.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 251 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHcuTSrXNll6FDDnkJ:-KLHcuTSrXNll6FDDnkK:328940337) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L251)):*
> ```Java
> 
> 
>         ChangeEvent(@NonNull String workspaceId, @Nullable String groupId, @NonNull String blockId,
> ```

javadoc

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 252 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHhMtJikyHgjV0qFLQ:-KLHhMtJikyHgjV0qFLR:504840414) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L252)):*
> ```Java
> 
>         ChangeEvent(@NonNull String workspaceId, @Nullable String groupId, @NonNull String blockId,
>                     @NonNull String fieldName, @NonNull @ChangeElement String elementChanged,
> ```

Should elementChanged be before fieldName?

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 308 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHhzPMognVUoziIOpO:-KLHhzPMognVUoziIOpP:-1714414601) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L308)):*
> ```Java
>         }
> 
>         public String getXml() {
> ```

Javadoc "Get the xml for the blocks that were created."

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 312 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHi5ihLKij6-Kjw5Zu:-KLHi5ihLKij6-Kjw5Zv:2146072478) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L312)):*
> ```Java
>         }
> 
>         public List<String> getIds() {
> ```

"Get the list of block ids that were created."

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 336 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHiBKzg_SvkRf7APSz:-KLHiBKzg_SvkRf7APT-:-543199854) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L336)):*
> ```Java
>         private final List<String> mIds;
> 
>         DeleteEvent(String workspaceId, String groupId, @Nullable String blockId, String oldXml, List<String> ids) {
> ```

nit length

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 341 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHiYbZ9gsSAXXUM4ff:-KLHiYbZ9gsSAXXUM4fg:-1594160511) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L341)):*
> ```Java
>             this.mIds = Collections.unmodifiableList(ids);
>         }
> ```

missing methods to get the oldxml and the ids

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 375 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHikp4-PNl-7QzA8hi:-KLHikp4-PNl-7QzA8hj:-1390483603) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L375)):*
> ```Java
> 
>         /**
>          * Creates a new
> ```

finish javadoc. Make it clear that the event should be created before the move and finished at the end of the move. To make it clearer, please pull out the initial setup into a recordOld method so it's obvious that there's two steps. This event is pretty different than the other events in use, so the more we can do to really make it clear the better.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Block.java, line 150 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHk1Y74NAL7egke6JR:-KLHk1Y74NAL7egke6JS:-327861014) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/model/Block.java#L150)):*
> ```Java
>                 connectedBlock.addAllBlockIds(outList);
>             }
>         }
> ```

Don't forget to recurse on nextConnection

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 122 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHtEGJCm_8pb1ZdUVx:-KLHtEGJCm_8pb1ZdUVy:1384717276) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L122)):*
> ```Java
>      *
>      * @param toSerialize A list of Blocks to serialize.
>      * @param os An OutputStream which to write them.
> ```

Please fix here too: "An OutputStream to write the blocks to."

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 135 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHt2kKcDWkNAy3p17K:-KLHt2kKcDWkNAy3p17L:-800404858) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L135)):*
> ```Java
>      *
>      * @param toSerialize A list of Blocks to serialize.
>      * @param writer An writer which to write them.
> ```

nit "A writer to  write the blocks to."

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 149 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHtLzZfWukYbw1SHEt:-KLHtLzZfWukYbw1SHEu:-1677837463) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L149)):*
> ```Java
>      *
>      * @param toSerialize A list of Blocks to serialize.
>      * @param os An OutputStream which to write them.
> ```

ditto.

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 200 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHtXaQf-q1f35OQCvZ:-KLHtXaQf-q1f35OQCv_:947737254) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L200)):*
> ```Java
>      * @throws BlocklySerializerException
>      */
>     public static String writeOneBlockToXmlString(Block toSerialize)
> ```

remove String from signature, it's implied by the return type. Add @return to the javadoc.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/335#-:-KLHtfY4M4vFc8X36dUf:1869906071)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 0 of 24 files reviewed at latest revision, 22 unresolved discussions.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/DevTestsActivity.java, line 83 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHXqb0ywEcIAJsaTuf:-KLI4TY1Is2ofOu_kyIm:-1019557266) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklydemo/src/main/java/com/google/blockly/android/demo/DevTestsActivity.java#L83)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> Why aren't you using id == R.id.log_events_menuitem? Then you could toggle 'item' directly instead of keeping a reference to it and you can remove the onCreateOptionsMenu override.
</details>
Moved to id. I saved there reference so that setLogEvents would only take a boolean (could be called without passing in the MenuItem reference).

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java, line 70 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLH_B4O7w_4AXCL8PGa:-KLIVOBKUBmrvxNkaeKD:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java#L70)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> s/Listener/EventsCallback
> 
> Add Javadoc
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java, line 77 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLH_OsBdGNT6xS7jxZi:-KLI7OKz5t5UgSbpAQ25:2029149825) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java#L77)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> "Called when a group of events are fired and at least one event is of a type specified by {@link #getTypesBitmask}. All of the events are not guaranteed to be of a specified type."
</details>
Rewrote.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java, line 837 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHZ0KJM_05oV5w1eB1:-KLRZIISjINd0ieeRJB9:1758359711) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java#L837)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> This should only be a create event if it's a new block, otherwise it's a move event with an old parent.
> 
> There are also cases where addRootBlock is called as part of a remove, please add a TODO to investigate.
</details>
Fixed.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 26 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHayjHviw0zbJn-2Xg:-KLRZHeMHzK1HOzgiHGZ:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L26)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> s/to/for
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 39 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHb42a4ULVGxx0vmH_:-KLRZhYPM3uYEcx4qfAJ:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L39)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> nit: >100 cols
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 92 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHbc5CfCFafR_06i5f:-KLRZlrke6ErgneW_BZ-:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L92)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> s/it/id
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 95 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHc3ozCvm1sa-gwqDN:-KLRZuI1NkjHW5MKPDTO:-604485601) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L95)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> These validate methods can probably be private.
</details>
They could, but there are probably a few cases where app code trying to configure Blockly might want them for their own configuration loading.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 172 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHcWIHTrHXDCzHlQDC:-KLRaJFRwUPdRzOVcAqk:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L172)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> javadoc
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 240 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHd15020ENc_nU03Fr:-KLRaTd1vHO2hHyc1sKl:191314312) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L240)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> Consider making the concrete event classes final. It'll discourage breaking compatibility with other events.
</details>
Done. Good point.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 251 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHcuTSrXNll6FDDnkJ:-KLRb3cndBL_xus86yQC:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L251)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> javadoc
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 252 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHhMtJikyHgjV0qFLQ:-KLRbIqYZbaFOFjll6E0:-263187594) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L252)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> Should elementChanged be before fieldName?
</details>
Moved it to the front.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 308 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHhzPMognVUoziIOpO:-KLZLIYvjOEZwmwft9mp:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L308)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> Javadoc "Get the xml for the blocks that were created."
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 312 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHi5ihLKij6-Kjw5Zu:-KLZLJzW1gjW8GAuECAh:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L312)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> "Get the list of block ids that were created."
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 336 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHiBKzg_SvkRf7APSz:-KLZLLsUDbcAuH05RhkB:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L336)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> nit length
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 341 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHiYbZ9gsSAXXUM4ff:-KLZLOYveT2rfEIcskF0:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L341)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> missing methods to get the oldxml and the ids
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 375 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHikp4-PNl-7QzA8hi:-KLZMkGVkKE06vJoXh0Q:-524315702) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L375)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> finish javadoc. Make it clear that the event should be created before the move and finished at the end of the move. To make it clearer, please pull out the initial setup into a recordOld method so it's obvious that there's two steps. This event is pretty different than the other events in use, so the more we can do to really make it clear the better.
</details>
Added documentation.  The current constructor and `recordNew()` mimics web blockly.  If we want to update that with a recordOld, I'll do the same here.  That said, there is no valid case of creating the event without immediately recording the old state (calling `recordOld()`), so I see the separation as extraneous.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Block.java, line 150 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHk1Y74NAL7egke6JR:-KLZsJsCosYPlAOipqqU:-368819667) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/model/Block.java#L150)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> Don't forget to recurse on nextConnection
</details>
Good catch.  Also clarified the behavior with shadow blocks in the comment.

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 122 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHtEGJCm_8pb1ZdUVx:-KLZsi0mOHAUAwMlEPDB:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L122)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> Please fix here too: "An OutputStream to write the blocks to."
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 135 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHt2kKcDWkNAy3p17K:-KLZsleZaxPuUnPVi81G:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L135)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> nit "A writer to  write the blocks to."
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 149 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHtLzZfWukYbw1SHEt:-KLZsyVxapMxfskkSvc7:-499146009) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L149)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> ditto.
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 200 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHtXaQf-q1f35OQCvZ:-KLZtXgLM-LA1Vhkrr0v:2003197610) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L200)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> remove String from signature, it's implied by the return type. Add @return to the javadoc.
</details>
Done and done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/335)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Review status: 0 of 24 files reviewed at latest revision, 26 unresolved discussions.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/DevTestsActivity.java, line 83 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHXqb0ywEcIAJsaTuf:-KLbLlGpStfri3hpY-w9:-921868830) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklydemo/src/main/java/com/google/blockly/android/demo/DevTestsActivity.java#L83)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary>
> Moved to id. I saved there reference so that setLogEvents would only take a boolean (could be called without passing in the MenuItem reference).
</details>
It might be better to have setLogEvents take the item instead. It can get the state itself and perform the update without needing the external reference.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 95 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHc3ozCvm1sa-gwqDN:-KLbNAVHnhltpIU_xkoe:1453345496) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L95)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary>
> They could, but there are probably a few cases where app code trying to configure Blockly might want them for their own configuration loading.
</details>
Do you have a specific example in mind? This seems like it's purely implementation details.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 375 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHikp4-PNl-7QzA8hi:-KLbQ3DR48ggxccBH_dH:-1639866545) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L375)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary>
> Added documentation.  The current constructor and `recordNew()` mimics web blockly.  If we want to update that with a recordOld, I'll do the same here.  That said, there is no valid case of creating the event without immediately recording the old state (calling `recordOld()`), so I see the separation as extraneous.
</details>
Good points. Please comment the expectation that recordNew is called at the new location before the event is fired.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 208 \[r2\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLbOkbel0b5BchW17rU:-KLbNrQevMUw37uTSuqq:1993593617) ([raw file](https://github.com/google/blockly-android/blob/29b5197bfcc283d301d183d1eb093723954f1d1d/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L208)):*
> ```Java
>     }
> 
>     protected BlocklyEvent(@EventType int typeId, JSONObject json) throws JSONException {
> ```

Javadoc

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 310 \[r2\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLbOomnUSzj74DOXRbn:-KLbOomnUSzj74DOXRbo:-1194106734) ([raw file](https://github.com/google/blockly-android/blob/29b5197bfcc283d301d183d1eb093723954f1d1d/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L310)):*
> ```Java
>          *
>          * @param elementChanged The {@link ChangeElement} identifying the aspect of the change.
>          * @param workspaceId The string id for the workspace.
> ```

Why does this version have workspaceId  and the field version not have groupId?

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 313 \[r2\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLbP5ZH_wYPQjj4B1-3:-KLbP5ZH_wYPQjj4B1-4:840291216) ([raw file](https://github.com/google/blockly-android/blob/29b5197bfcc283d301d183d1eb093723954f1d1d/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L313)):*
> ```Java
>          * @param groupId The string id for the event group.
>          * @param block The block containing the change.
>          * @param oldValue The original field value.
> ```

This version doesn't support changes to field, so these values can be field values.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Block.java, line 60 \[r2\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLbRGD3zVmaSckcFvId:-KLbRGD4f6O7eCvA398Q:305545341) ([raw file](https://github.com/google/blockly-android/blob/29b5197bfcc283d301d183d1eb093723954f1d1d/blocklylib-core/src/main/java/com/google/blockly/model/Block.java#L60)):*
> ```Java
> 
>     // These values can be changed after creating the block
>     private Workspace mWorkspace;
> ```

I'm not sure it's worth the extra logic and reference just to avoid passing the workspace id in when creating events... Is there some other reason this is necessary?

---

*[blocklylib-core/src/main/java/com/google/blockly/model/FieldDropdown.java, line 142 \[r2\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLbRhoffT_cVuAbuJ3K:-KLbRhoffT_cVuAbuJ3L:-1317828125) ([raw file](https://github.com/google/blockly-android/blob/f75f5717b299dc156ed40af178648225fa71a302/blocklylib-core/src/main/java/com/google/blockly/model/FieldDropdown.java#L142)):*
> ```Java
>      * @return The value of the currently selected option.
>      */
>     public String getSelectedValue() {
> ```

Why was this deleted? It's the pair of setSelectedValue

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/335)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 0 of 27 files reviewed at latest revision, 26 unresolved discussions.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/DevTestsActivity.java, line 83 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHXqb0ywEcIAJsaTuf:-KLc47lK7nPFX5qs2w1H:1634841178) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklydemo/src/main/java/com/google/blockly/android/demo/DevTestsActivity.java#L83)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> It might be better to have setLogEvents take the item instead. It can get the state itself and perform the update without needing the external reference.
</details>
Fixed per offline conversation.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 95 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHc3ozCvm1sa-gwqDN:-KLc5Y6jkRZyCiWiAWo-:377045410) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L95)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> Do you have a specific example in mind? This seems like it's purely implementation details.
</details>
With the new constructors, it make more sense to making private, as you suggested.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 375 \[r1\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLHikp4-PNl-7QzA8hi:-KLc9LWkO8XeSLx5-9L9:1925452848) ([raw file](https://github.com/google/blockly-android/blob/b85b51df395dae80c3e3cc31d8e5a594e328b57b/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L375)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> Good points. Please comment the expectation that recordNew is called at the new location before the event is fired.
</details>
Added a sentence to that effect.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 208 \[r2\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLbOkbel0b5BchW17rU:-KLc6atvwXRyc0cg4dwv:-499146009) ([raw file](https://github.com/google/blockly-android/blob/29b5197bfcc283d301d183d1eb093723954f1d1d/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L208)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> Javadoc
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 310 \[r2\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLbOomnUSzj74DOXRbn:-KLc7VI9btwZgmaTFRVN:852050802) ([raw file](https://github.com/google/blockly-android/blob/29b5197bfcc283d301d183d1eb093723954f1d1d/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L310)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> Why does this version have workspaceId  and the field version not have groupId?
</details>
Replace workspaceId with block.getWorkspaceId().

Group id is set to all members of a group just before the group is set out to BlocklyController.EventCallbacks, unlike web Blockly which sets and unsets a static current group id.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 313 \[r2\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLbP5ZH_wYPQjj4B1-3:-KLc829MEHU2G1Sn0F_l:-1196653477) ([raw file](https://github.com/google/blockly-android/blob/29b5197bfcc283d301d183d1eb093723954f1d1d/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L313)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> This version doesn't support changes to field, so these values can be field values.
</details>
Correct @param docs.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Block.java, line 60 \[r2\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLbRGD3zVmaSckcFvId:-KLcCP0f9DuJIawn63vr:-1962810618) ([raw file](https://github.com/google/blockly-android/blob/29b5197bfcc283d301d183d1eb093723954f1d1d/blocklylib-core/src/main/java/com/google/blockly/model/Block.java#L60)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> I'm not sure it's worth the extra logic and reference just to avoid passing the workspace id in when creating events... Is there some other reason this is necessary?
</details>
I was on the fence when I wrote it, but it felt cleaner.  Removed, now.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/FieldDropdown.java, line 142 \[r2\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KLbRhoffT_cVuAbuJ3K:-KLc9pUBj6x3jZ30QMnY:1357570301) ([raw file](https://github.com/google/blockly-android/blob/f75f5717b299dc156ed40af178648225fa71a302/blocklylib-core/src/main/java/com/google/blockly/model/FieldDropdown.java#L142)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> Why was this deleted? It's the pair of setSelectedValue
</details>
Good point.  Removed because it was identical to getSerailizedValue(), but not just have that delegate to this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/335)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Review status: 0 of 27 files reviewed at latest revision, 7 unresolved discussions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 352 \[r3\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KM0-4SdoI8I6z8i09NH:-KM0-4SenkHYNKwuqA28:-1802169647) ([raw file](https://github.com/google/blockly-android/blob/415d094cfb4cc776c4d17b9fc70921618922f7a8/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L352)):*
> ```Java
>          *              null.
>          * @param oldValue The original value.
>          * @param newValue The original value.
> ```

The new value.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Block.java, line 683 \[r3\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KM0-aNk_VuVt_Tf0uni:-KM0-aNk_VuVt_Tf0unj:1445946052) ([raw file](https://github.com/google/blockly-android/blob/415d094cfb4cc776c4d17b9fc70921618922f7a8/blocklylib-core/src/main/java/com/google/blockly/model/Block.java#L683)):*
> ```Java
> 
>     /**
> <<<<<<< HEAD
> ```

bad rebase

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 241 \[r3\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KM07M6fh3gID7nKYK63:-KM07M6gkx5mn_jmPHkl:-860247281) ([raw file](https://github.com/google/blockly-android/blob/415d094cfb4cc776c4d17b9fc70921618922f7a8/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L241)):*
> ```Java
>      *
>      * @param inStream The input stream to read blocks from. Maybe null.
>      * @param inString The xml string to read blocks from if {@code is} is null.
> ```

{@code inStream}

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/335)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Once comments are fixed <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 0 of 27 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/335#-:-KM07e1rR-gTnhluQGY5:-1696123560)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 0 of 27 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java, line 352 \[r3\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KM0-4SdoI8I6z8i09NH:-KM07dbZPae_0ckGbksf:-499146009) ([raw file](https://github.com/google/blockly-android/blob/415d094cfb4cc776c4d17b9fc70921618922f7a8/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyEvent.java#L352)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> The new value.
</details>
Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Block.java, line 683 \[r3\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KM0-aNk_VuVt_Tf0uni:-KM07g7ZMZYb0wcekjl5:1758359711) ([raw file](https://github.com/google/blockly-android/blob/415d094cfb4cc776c4d17b9fc70921618922f7a8/blocklylib-core/src/main/java/com/google/blockly/model/Block.java#L683)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> bad rebase
</details>
Fixed.

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 241 \[r3\]](https://reviewable.io:443/reviews/google/blockly-android/335#-KM07M6fh3gID7nKYK63:-KM082xEixBF6TnYspFT:-499146009) ([raw file](https://github.com/google/blockly-android/blob/415d094cfb4cc776c4d17b9fc70921618922f7a8/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L241)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary>
> {@code inStream}
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/335)*
<!-- Sent from Reviewable.io -->
